clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java

```diff
@@ -136,27 +369,160 @@ public ConfigDef withClientSaslSupport() {
                 value = key.defaultValue;
-            if (key.validator != null)
+            }
+            if (value != null && key.validator != null) {
                 key.validator.ensureValid(key.name, value);
-            values.put(key.name, value);
+            }
+            // only put the value that actually parsed.
+            if (value != null) {
+                values.put(key.name, value);
+            }
         }
         return values;
     }
 
     /**
+     * Validate the provided connector configuration values with the configuration definition
+     * @param props the provided configuration values
+     * @return List of Config, each Config contains the updated configuration information given
+     * the current configuration values.
+     */
+    public List<ConfigValue> validate(Map<String, String> props) {
+        List<String> configsWithNoParent = getConfigsWithNoParent();
+        Map<String, ConfigValue> configValues = new HashMap<>();
+        for (String name: configKeys.keySet()) {
+            configValues.put(name, new ConfigValue(name));
+        }
+        try {
+            ensureDependentsDefined();
+        } catch (ConfigException e) {
+            configValues.put(undefinedConfigKey, new ConfigValue(undefinedConfigKey));
+        }
+        
+        for (String name: configsWithNoParent) {
+            validate(name, props, configValues);
+        }
+        return new LinkedList<>(configValues.values());
+    }
+
+    private void ensureDependentsDefined() {
+        for (String configName: configKeys.keySet()) {
+            ConfigKey configKey = configKeys.get(configName);
+            List<String> dependents = configKey.dependents;
+            for (String dependent: dependents) {
+                if (!configKeys.containsKey(dependent)) {
+                    undefinedConfigKey = dependent;
+                    throw new ConfigException(dependent + " is  an dependent of " + configName + ". However, it is not defined.");
+                }
+            }
+        }
+    }
+
+    private List<String> getConfigsWithNoParent() {
+        Set<String> configs = configKeys.keySet();
+        Set<String> configsWithParent = new HashSet<>();
+
+        for (ConfigKey configKey: configKeys.values()) {
+            List<String> dependents = configKey.dependents;
+            configsWithParent.addAll(dependents);
+        }
+
+        List<String> configsWithNoParent = new LinkedList<>();
+        for (String config: configs) {
+            if (!configsWithParent.contains(config)) {
+                configsWithNoParent.add(config);
+            }
+        }
+        return configsWithNoParent;
+    }
+
+    private void validate(String name, Map<String, String> props, Map<String, ConfigValue> configs) {
+        ConfigKey key = configKeys.get(name);
+        ConfigValue config = configs.get(name);
+
+        Object value;
+        if (props.containsKey(key.name)) {
```

Comment:
Isn't this just duplicated from the `parse` method? Can they not be combined because of the difference in error handling?